### PR TITLE
Merge remaining dcrutil code into a dcrd package.

### PR DESCRIPTION
**Pull request by jrick**
This merge commit adds the following code from the
github.com/decred/dcrutil package into a new
github.com/decred/dcrd/dcrutil package:

* Address handling
* Amount type
* AppDataDir func
* bitflags functions
* Block wrapper type
* Hash160 func
* Tx wrapper type
* WIF type

as well as all tests for this code.

The old github.com/decred/dcrutil/hdkeychain package has also been
merged and moved to github.com/decred/dcrd/dcrutil/hdkeychain.

dcrd packages have been updated to use the new packages and the dep
files have been updated for this change.

### CODE REVIEW

File: dcrutil/README.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
+dcrutil
+=======
+
+
+[![Build Status](http://img.shields.io/travis/decred/dcrutil.svg)](https://travis-ci.org/decred/dcrutil)
```
Comment (davecgh) [inline dcrutil/README.md]: This is incorrect.  Needs to be `[http://img.shields.io/travis/decred/dcrd.svg](https://travis-ci.org/decred/dcrd)`

File: dcrutil/README.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+[![Coverage Status](http://img.shields.io/coveralls/decred/dcrutil.svg)](https://coveralls.io/r/decred/dcrutil?branch=master)
```
Comment (davecgh) [inline dcrutil/README.md]: This needs to be removed since coveralls isn't setup for the overall dcrd repo.

File: dcrutil/README.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+[![ISC License](http://img.shields.io/badge/license-ISC-blue.svg)](http://copyfree.org)
+[![GoDoc](http://img.shields.io/badge/godoc-reference-blue.svg)](http://godoc.org/github.com/decred/dcrutil)
```
Comment (davecgh) [inline dcrutil/README.md]: Incorrect path.

File: dcrutil/README.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+This package was developed for dcrd, a full-node implementation of Decred which
+is under active development by Company 0.  Although it was primarily written for
+dcrd, this package has intentionally been designed so it can be used as a
+standalone package for any projects needing the functionality provided.
+
+## Installation and Updating
+
+```bash
+$ go get -u github.com/decred/dcrutil
```
Comment (davecgh) [inline dcrutil/README.md]: Incorrect path.

File: dcrutil/README.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+```
+
+## Docker
```
Comment (davecgh) [inline dcrutil/README.md]: This section no longer applies.  Should be removed.

File: dcrutil/address_test.go [hunk 6]
```diff
@@ -0,0 +1,579 @@
...
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"reflect"
+	"testing"
+
+	"github.com/decred/dcrd/chaincfg"
+	"github.com/decred/dcrd/chaincfg/chainec"
+	"github.com/decred/dcrd/wire"
+	"github.com/decred/dcrutil"
```
Comment (davecgh) [inline dcrutil/address_test.go]: Incorrect path.

File: dcrutil/amount_test.go [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+
+package dcrutil_test
+
+import (
+	"math"
+	"reflect"
+	"sort"
+	"testing"
+
+	. "github.com/decred/dcrutil"
```
Comment (davecgh) [inline dcrutil/amount_test.go]: Incorrect path.

File: dcrutil/appdata_test.go [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+import (
+	"os"
+	"os/user"
+	"path/filepath"
+	"runtime"
+	"testing"
+	"unicode"
+
+	"github.com/decred/dcrutil"
```
Comment (davecgh) [inline dcrutil/appdata_test.go]: Incorrect path.

File: dcrutil/block_test.go [hunk 9]
```diff
@@ -0,0 +1,562 @@
...
+	"bytes"
+	"io"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/davecgh/go-spew/spew"
+	"github.com/decred/dcrd/chaincfg/chainhash"
+	"github.com/decred/dcrd/wire"
+	"github.com/decred/dcrutil"
```
Comment (davecgh) [inline dcrutil/block_test.go]: Incorrect path.

File: dcrutil/example_test.go [hunk 10]
```diff
@@ -0,0 +1,76 @@
+package dcrutil_test
+
+import (
+	"fmt"
+	"math"
+
+	"github.com/decred/dcrutil"
```
Comment (davecgh) [inline dcrutil/example_test.go]: Incorrect path.

File: dcrutil/tx_test.go [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+
+import (
+	"bytes"
+	"io"
+	"reflect"
+	"testing"
+
+	"github.com/davecgh/go-spew/spew"
+	"github.com/decred/dcrd/chaincfg/chainhash"
+	"github.com/decred/dcrutil"
```
Comment (davecgh) [inline dcrutil/tx_test.go]: Incorrect path.

File: dcrutil/wif_test.go [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+// license that can be found in the LICENSE file.
+
+package dcrutil_test
+
+import (
+	"testing"
+
+	"github.com/decred/dcrd/chaincfg"
+	"github.com/decred/dcrd/chaincfg/chainec"
+	. "github.com/decred/dcrutil"
```
Comment (davecgh) [inline dcrutil/wif_test.go]: Incorrect path.

**Comment by jrick** [comment]
updated

**Comment by jrick** [comment]
i think this may technically break the rpctest integration test which isn't run by travis, since it is using the old hdkeychain still in the dcrutil repo and it was not updated for the address code moved to dcrd.  Since this is a chicken-and-egg problem I'm intentionally ignoring it until hdkeychain is moved into dcrd.

File: dcrutil/README.md [hunk 13]
```diff
@@ -0,0 +1,26 @@
+dcrutil
+=======
+
+
+[![Build Status](http://img.shields.io/travis/decred/dcrutil.svg)](https://travis-ci.org/decred/dcrd)
```
Comment (davecgh) [inline dcrutil/README.md]: This is still incorrect.   Needs to be `[http://img.shields.io/travis/decred/dcrd.svg](https://travis-ci.org/decred/dcrd)`.

Comment (jrick) [reply dcrutil/README.md]: sorry, fixing

**Comment by jrick** [comment]
I would definitely prefer to keep all history but I merged it this way since it (like bloom, and like hdkeychain will be) is only a subset of all of the code from that repo.

**Comment by davecgh** [comment]
It should be pretty easy to do.  Here is the process I followed for the rpcclient and others in the past:

* Add a remote for `dcrutil` and fetch it
* Creat a new `import_dcrutil` branch against `dcrutil/master`
* Use `git rm` and `git mv` to remove unneeded files/packages and move the remaining relevant files to the `dcrutil` directory
* Modify the import paths for the `dcrutil` code only (not changes to dcrd since that code is not even in the local tree at this point due to being in an unrelated history), `dcrutil/README.md`, etc.
* Commit those changes `Import dcrutil repo into dcrutil directory.`
* Create a new `merge_dcrutil` branch against `origin/master` (aka dcrd)
* Perform an unrelated history merge (e.g `git merge --allow-unrelated-histories import_dcrutil`) with a title of something like `dcrutil: Merge dcrutil repo`.
* Modify all of the dcrd code and README references to point to the new path
* Amend the merge commit to include all of those changes
* Submit the PR

**Comment by jrick** [comment]
i know how to do the merge.  my point was more that this code isn't any more special than any of the code that has been imported or will be imported later.  why would it get history when the rest doesn't?

**Comment by davecgh** [comment]
All of the other code has does have the history however.  We do upstream merges with btcd, the rpcclient merge we just did, if you notice, had all history, etc.

I would not have approved the certgen PR that went in earlier either for the same reason and I was actually quite upset it was merged to be honest.

**Comment by jrick** [comment]
and bloom? you approved *and merged* that commit.

**Comment by davecgh** [comment]
Yep, and that was an oversight on my part.  I realized it several hours after I had already merged it and it was too late by that point.

**Comment by jrick** [comment]
The timeline doesn't match.  If you would not have approved certgen for the exact same reason as discussed for bloom, then you would have not approved bloom since it was a case of the exact same thing.

I will merge it as requested since I need these changes but I don't appreciate the history rewriting double standards.

**Comment by jrick** [comment]
Merged as requested.

The travis failure appears to be a bug with vet that @dajohi had noticed earlier in the dcrutil repo.  I don't know how we plan to handle that here.

**Comment by davecgh** [comment]
Thanks.  Reviewing again.

**Pull request merged**